tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ConcreteFunction.java

```diff
@@ -154,78 +178,191 @@ public Signature signature() {
+    for (int i = 0; i < outputNames.size(); i++) {
+      String outputName = outputNames.get(i);
 
-    } catch (Exception e) {
-      // Release tensors before throwing exception
-      for (Tensor t : resultTensors) {
-        t.close();
+      if (i > outputList.size()) {
+        throw new IllegalStateException("Somehow, not all required outputs were returned from the function");
       }
-      throw e;
+
+      Operand<?> output = outputList.get(i);
+      namedOutputs.put(outputName, output);
     }
+
+    return Collections.unmodifiableMap(namedOutputs);
   }
 
   /**
-   * Invokes a function with a single input and output.
+   * Calls the function in an execution environment, adding it's graph as a function if it isn't already present. Only
+   * works for functions with a single input and output.
    *
-   * <p>Caller is responsible for closing all Tensors.
-   *
-   * @param tensor input tensor
-   * @return output tensor
-   * @throws IllegalArgumentException if there are multiple input or output parameters defined
-   *                                  in the function
+   * @param scope the scope to call the function in
+   * @param argument the argument to the call
+   * @return the output of the function
    */
-  public Tensor call(Tensor tensor) throws IllegalArgumentException {
+  public Operand<?> call(Scope scope, Operand<?> argument) {
     final SignatureDef signatureDef = signature.asSignatureDef();
 
     if (signatureDef.getInputsCount() != 1) {
       throw new IllegalArgumentException(
-        String.format("Function [%s] requires multiple inputs", signatureDef.getMethodName()));
+          String.format("Function [%s] requires multiple inputs", signatureDef.getMethodName()));
     }
-    String inputNodeName = signatureDef.getInputsMap().values().iterator().next().getName();
+    String inputName = signatureDef.getInputsMap().keySet().iterator().next();
 
     if (signatureDef.getOutputsCount() != 1) {
       throw new IllegalArgumentException(
-        String.format("Function [%s] has multiple outputs", signatureDef.getMethodName()));
+          String.format("Function [%s] has multiple outputs", signatureDef.getMethodName()));
+    }
+    String outputName = signatureDef.getOutputsMap().keySet().iterator().next();
+
+    Map<String, Operand<?>> inputMap = new LinkedHashMap<>();
+    inputMap.put(inputName, argument);
+
+    return call(scope, inputMap).get(outputName);
+  }
+
+  /**
+   * Invokes a function using the default eager session.
+   *
+   * <p>Caller is responsible for closing all Tensors.
+   *
+   * @param arguments list of tensors to pass in input to the function, mapped by their signature name
+   * @return output tensors resulting from the execution of the function, mapped by their signature name
+   */
+  public Map<String, Tensor> call(Map<String, Tensor> arguments)
+      throws IllegalArgumentException {
+    //FIXME need to manage input/output operand lifetimes
```

Comment:
Yeah... it's still not clear how "dangerous" the default eager session is, can we wait a bit before promoting this specific endpoint or its required for your work?

But that's the part I'm still not sure about, see my [previous comment](https://github.com/tensorflow/java/pull/233/files#r608277505). When the function is issued or backed by a `Session` and not a `NativeFunction`, we should just run that session internally like it was done before. So this is not supported at all anymore? It is required for executing the main graph of a saved model built with TF2.x in Python.

Reply:
ConcreteFunction isn't session backed any more, that's what the new `SavedFunction` is for.

I think using the default session is fine for now, the fixme was for once we have resource scopes.  I could spin up a new eager session each time, but that seems expensive.